File: go_backend/services/alerts/alerts.go
package alerts

import (
	"log"
	"sync"
)

var (
	notifiers []Notifier
	enabled   bool
	mu        sync.RWMutex
)

// Init initializes the alerting system with enabled notifiers
func Init(isEnabled bool, enabledNotifiers ...Notifier) {
	mu.Lock()
	defer mu.Unlock()
	enabled = isEnabled
	notifiers = enabledNotifiers
}

// SetEnabled toggles the alerting system
func SetEnabled(isEnabled bool) {
	mu.Lock()
	defer mu.Unlock()
	enabled = isEnabled
}

// Info shoots an info alert to all enabled channels
func Info(message string) {
	send(LevelInfo, message)
}

// Warn shoots a warning alert to all enabled channels
func Warn(message string) {
	send(LevelWarn, message)
}

// Error shoots an error alert to all enabled channels
func Error(message string) {
	send(LevelError, message)
}

func send(level Level, message string) {
	mu.RLock()
	defer mu.RUnlock()

	if !enabled {
		return
	}

	for _, n := range notifiers {
		go func(notifier Notifier) {
			if err := notifier.Send(level, message); err != nil {
				log.Printf("failed to send alert via notifier: %v", err)
			}
		}(n)
	}
}

File: node_backend/src/services/alerts/alerts.ts
// services/alerts/alerts.ts — Equivalent to go_backend/services/alerts/alerts.go
import { Notifier, Level, LevelInfo, LevelWarn, LevelError } from './notifier';

let notifiers: Notifier[] = [];
let enabled: boolean = false;

/**
 * Initialize the alerting system with enabled notifiers.
 * Mirrors the Go Init function.
 */
export function init(isEnabled: boolean, ...enabledNotifiers: Notifier[]): void {
    enabled = isEnabled;
    notifiers = enabledNotifiers;
}

/**
 * Toggle the alerting system on/off.
 */
export function setEnabled(isEnabled: boolean): void {
    enabled = isEnabled;
}

/**
 * Send an info alert to all enabled channels.
 */
export function info(message: string): void {
    send(LevelInfo, message);
}

/**
 * Send a warning alert to all enabled channels.
 */
export function warn(message: string): void {
    send(LevelWarn, message);
}

/**
 * Send an error alert to all enabled channels.
 */
export function error(message: string): void {
    send(LevelError, message);
}

function send(level: Level, message: string): void {
    if (!enabled) {
        return;
    }

    for (const notifier of notifiers) {
        // Fire and forget — mirrors Go's goroutine dispatch
        notifier.send(level, message).catch((err) => {
            console.error(`Failed to send alert via notifier: ${err}`);
        });
    }
}

export { LevelInfo, LevelWarn, LevelError };
